cpp/include/kvikio/shim/libcurl.hpp

```diff
@@ -0,0 +1,260 @@
+
+    // We always set CURLOPT_ERRORBUFFER to get better error messages.
+    setopt(CURLOPT_ERRORBUFFER, _errbuf);
+
+    // Make curl_easy_perform() fail when receiving HTTP code errors.
+    setopt(CURLOPT_FAILONERROR, 1L);
+  }
+  ~CurlHandle() noexcept { LibCurl::instance().retain_handle(std::move(_handle)); }
+
+  /**
+   * @brief CurlHandle support is not movable or copyable.
+   */
+  CurlHandle(CurlHandle const&)            = delete;
+  CurlHandle& operator=(CurlHandle const&) = delete;
+  CurlHandle(CurlHandle&& o)               = delete;
+  CurlHandle& operator=(CurlHandle&& o)    = delete;
+
+  /**
+   * @brief Get the underlying curl easy handle pointer.
+   */
+  CURL* handle() noexcept { return _handle.get(); }
+
+  /**
+   * @brief Set option for the curl handle.
+   *
+   * See <https://curl.se/libcurl/c/curl_easy_setopt.html> for available options.
+   *
+   * @tparam VAL The type of the value.
+   * @param option The curl option to set.
+   */
+  template <typename VAL>
+  void setopt(CURLoption option, VAL value)
+  {
+    CURLcode err = curl_easy_setopt(handle(), option, value);
+    if (err != CURLE_OK) {
+      std::stringstream ss;
+      ss << "curl_easy_setopt() error near " << _source_file << ":" << _source_line;
+      ss << "(" << curl_easy_strerror(err) << ")";
+      throw std::runtime_error(ss.str());
+    }
+  }
+
+  /**
+   * @brief Perform a blocking network transfer using previously set options.
+   *
+   * See <https://curl.se/libcurl/c/curl_easy_perform.html>.
+   */
+  void perform()
+  {
+    // Perform the curl operation and check for errors.
+    CURLcode err = curl_easy_perform(handle());
+    if (err != CURLE_OK) {
+      std::string msg(_errbuf);
+      std::stringstream ss;
+      ss << "curl_easy_perform() error near " << _source_file << ":" << _source_line;
+      if (msg.empty()) {
+        ss << "(" << curl_easy_strerror(err) << ")";
+      } else {
+        ss << "(" << msg << ")";
+      }
+      throw std::runtime_error(ss.str());
+    }
+  }
+
+  /**
+   * @brief Extract information from a curl handle.
+   *
+   * See <https://curl.se/libcurl/c/curl_easy_getinfo.html> for available options.
+   *
+   * @tparam OUTPUT The type of the output.
+   * @param output The output, which is used as-is: `curl_easy_getinfo(..., output)`.
+   */
+  template <typename OUTPUT>
+  void getinfo(CURLINFO info, OUTPUT output)
+  {
+    CURLcode err = curl_easy_getinfo(handle(), info, output);
+    if (err != CURLE_OK) {
+      std::stringstream ss;
+      ss << "curl_easy_getinfo() error near " << _source_file << ":" << _source_line;
+      ss << "(" << curl_easy_strerror(err) << ")";
+      throw std::runtime_error(ss.str());
+    }
+  }
+};
+
+/**
+ * @brief Create a new curl handle.
+ *
+ * @returns A `kvikio::CurlHandle` instance ready to be used.
+ */
+#define create_curl_handle() \
+  kvikio::CurlHandle(kvikio::LibCurl::instance().get_handle(), __FILE__, KVIKIO_STRINGIFY(__LINE__))
```

Comment:
*note*: This macro is only needed to get access to `__FILE__` and `__LINE__`? Note that once we go to C++20 we can use [`std::source_location`](https://en.cppreference.com/w/cpp/utility/source_location)